output/cloud/expv2/output_test.go

```diff
@@ -278,3 +280,87 @@ func TestOutputStopWithTestError(t *testing.T) {
 	require.NoError(t, o.Start())
 	require.NoError(t, o.StopWithTestError(errors.New("an error")))
 }
+
+func TestOutputFlushMetricsConcurrently(t *testing.T) {
+	t.Parallel()
+
+	done := make(chan struct{})
+
+	// It blocks on the first request so it asserts that the flush
+	// operations continues concurrently if one more tick is sent in the meantime.
+	//
+	// The second request unblocks.
+	var requestsCount int64
+	flusherMock := func() {
+		updated := atomic.AddInt64(&requestsCount, 1)
+		if updated == 2 {
+			close(done)
+			return
+		}
+		<-done
+	}
+
+	o := Output{logger: testutils.NewLogger(t)}
+	o.config.MetricPushConcurrency = null.IntFrom(2)
+	o.config.MetricPushInterval = types.NullDurationFrom(1) // loop
+	o.flushing = flusherFunc(flusherMock)
+	o.runFlushWorkers()
+
+	select {
+	case <-time.After(5 * time.Second):
+		t.Error("timed out")
+	case <-done:
+		assert.NotZero(t, atomic.LoadInt64(&requestsCount))
+	}
+}
+
+func TestOutputFlushWorkersStop(t *testing.T) {
+	t.Parallel()
+
+	o := Output{
+		logger: testutils.NewLogger(t),
+		stop:   make(chan struct{}),
+	}
+	o.config.MetricPushInterval = types.NullDurationFrom(1 * time.Millisecond)
+
+	once := sync.Once{}
+	flusherMock := func() {
+		// it asserts that flushers are set and the flush is invoked
+		once.Do(func() { close(o.stop) })
+	}
+
+	o.flushing = flusherFunc(flusherMock)
+	o.runFlushWorkers()
+
+	// it asserts that all flushers exit
+	o.wg.Wait()
```

Comment:
Could we add a timeout to both this test and the one below, to ensure that if this doesn't work, the test doesn't hang indefinitely?